### PR TITLE
feat(v5): New refactorings: wrap with useMemo, useCallback useEffect, rename state veriable

### PR DESCRIPTION
**Pull request by borislit**
BREAKING CHANGE: Hooks support will now be activated automatically based on React version (16.8 and up) instead of a flag

### CODE REVIEW

File: src/modules/rename-state.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        (pathEnd.line === end.line && pathEnd.column === end.character))
+
+}
+
+export function isStateVariable(text) {
+    try {
+        const allAST = codeToAst(allText());
+        const containerPath = findPathInContext(allAST, text);
+        const variableDeclarationParent = containerPath.scope.bindings[containerPath.node.name].path.parentPath;
+        return variableDeclarationParent && variableDeclarationParent.node.declarations[0].init.callee.name === 'useState'
```
Comment (loonydev) [inline src/modules/rename-state.ts:26]: Codacy Issue found: ['magic numbers' are not allowed: 0](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

**Comment by loonydev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 72
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229&bid=18143887)

File: src/modules/rename-state.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+export async function renameState() {
+    const selectedStateVariable = selectedText();
+    const varName = await showInputBox(null, 'Name of the state variable');
+
+    const allAST = codeToAst(allText());
+    const containerPath = findPathInContext(allAST, selectedStateVariable);
+    const variableDeclarationParent = containerPath.findParent(parent => t.isVariableDeclaration(parent));
+    if (variableDeclarationParent && variableDeclarationParent.node.declarations[0].init.callee.name === 'useState') {
```
Comment (loonydev) [inline src/modules/rename-state.ts:39]: Codacy Issue found: ['magic numbers' are not allowed: 0](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/wrap-with-usecallback.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    try {
+        const allAST = codeToAst(allText());
+        const containerPath = findSelectedFunctionInRange(allAST, selectedTextStart(), selectedTextEnd());
+        return isFunction(containerPath) && t.isBlockStatement(containerPath.parent)
+    } catch (e) {
+        return false;
+    }
+}
+
+function functionDeclartionToExpression(declarations: t.FunctionDeclaration) {
```
Comment (loonydev) [inline src/modules/wrap-with-usecallback.ts:44]: Codacy Issue found: [expected call-signature: 'functionDeclartionToExpression' to have a typedef](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/rename-state.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        const allAST = codeToAst(allText());
+        const containerPath = findPathInContext(allAST, text);
+        const variableDeclarationParent = containerPath.scope.bindings[containerPath.node.name].path.parentPath;
+        return variableDeclarationParent && variableDeclarationParent.node.declarations[0].init.callee.name === 'useState'
+    } catch (e) {
+        return false;
+    }
+}
+
+export async function renameState() {
```
Comment (loonydev) [inline src/modules/rename-state.ts:32]: Codacy Issue found: [non-arrow functions are forbidden](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/ast-helpers.ts [hunk 5]
```diff
@@ -11,3 +13,29 @@ export function getReactImportReference(ast): t.ImportDeclaration {
 export function isExportedDeclaration(ast) {
   return t.isExportNamedDeclaration(ast) || t.isExportDefaultDeclaration(ast);
 }
+
+export function findPathInContext(ast, identifierName) {
```
Comment (loonydev) [inline src/ast-helpers.ts:17]: Codacy Issue found: [expected parameter: 'identifierName' to have a typedef](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/rename-state.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import { selectedText, allText, selectedTextStart, selectedTextEnd, activeFileName, activeEditor, showInputBox } from "../editor";
+import traverse from "@babel/traverse";
+import { transformFromAst } from "@babel/core";
+import { window, Position, Range } from "vscode";
+import { persistFileSystemChanges, replaceTextInFile } from "../file-system";
+import { capitalizeFirstLetter } from "../utils";
+import * as t from "@babel/types";
+import { findPathInContext } from "../ast-helpers";
+
+function isPathOnLines(path, start, end) {
```
Comment (loonydev) [inline src/modules/rename-state.ts:11]: Codacy Issue found: [expected parameter: 'end' to have a typedef](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/rename-state.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+export async function renameState() {
+    const selectedStateVariable = selectedText();
+    const varName = await showInputBox(null, 'Name of the state variable');
+
+    const allAST = codeToAst(allText());
+    const containerPath = findPathInContext(allAST, selectedStateVariable);
+    const variableDeclarationParent = containerPath.findParent(parent => t.isVariableDeclaration(parent));
+    if (variableDeclarationParent && variableDeclarationParent.node.declarations[0].init.callee.name === 'useState') {
+        containerPath.scope.bindings[selectedStateVariable].path.parentPath.get('declarations.0.id.elements.0').scope.rename(selectedStateVariable, varName);
```
Comment (loonydev) [inline src/modules/rename-state.ts:40]: Codacy Issue found: [Exceeds maximum line length of 120](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/wrap-with-usecallback.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+export async function wrapWithUseCallback() {
+    const snippet = selectedText();
+    const ast = codeToAst(snippet);
+    const callackWrapper = buildUseCallbackHook({
+        CALLBACK: t.isFunctionDeclaration(ast.program.body[0]) ? functionDeclartionToExpression(ast.program.body[0] as t.FunctionDeclaration) : (ast.program.body[0] as t.VariableDeclaration).declarations[0].init
+    });
+
+    const callbackId = t.isFunctionDeclaration(ast.program.body[0]) ? (ast.program.body[0] as t.FunctionDeclaration).id : (ast.program.body[0] as t.VariableDeclaration).declarations[0].id;
```
Comment (loonydev) [inline src/modules/wrap-with-usecallback.ts:56]: Codacy Issue found: [Exceeds maximum line length of 120](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/extension.ts [hunk 9]
```diff
@@ -19,6 +23,34 @@ export class CompleteActionProvider implements vscode.CodeActionProvider {
...
+      return [{
+        command: 'extension.glean.react.rename-state-hook',
+        title: 'Rename State'
+      }];
+    }
+
+    if (isVariableDeclarationWithNonFunctionInit(text)) {
+      return [{
+        command: 'extension.glean.react.wrap-with-usememo',
+        title: 'Wrap with useMemo'
```
Comment (loonydev) [inline src/extension.ts:36]: Codacy Issue found: [Missing trailing comma](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/rename-state.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        containerPath.scope.bindings[varName].path.parentPath.get('declarations.0.id.elements.1').scope.rename(`set${capitalizeFirstLetter(selectedStateVariable)}`, `set${capitalizeFirstLetter(varName)}`);
+
+        const processedJSX = transformFromAst(allAST).code;
+        const endLine = activeEditor().document.lineAt(activeEditor().document.lineCount - 1).range;
+        const change = replaceTextInFile(
+            processedJSX,
+            new Position(0, 0),
+            endLine.end,
+            activeFileName()
```
Comment (loonydev) [inline src/modules/rename-state.ts:49]: Codacy Issue found: [Missing trailing comma](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/wrap-with-usememo.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    try {
+        const ast = codeToAst(text);
+        if (ast.program.body.length > 1) return false;
+        return t.isVariableDeclaration(ast.program.body[0]) && !isFunction((ast.program.body[0] as t.VariableDeclaration).declarations[0].init);
+    } catch (e) {
+        return false;
+    }
+}
+
+export async function wrapWithUseMemo() {
```
Comment (loonydev) [inline src/modules/wrap-with-usememo.ts:25]: Codacy Issue found: [expected call-signature: 'wrapWithUseMemo' to have a typedef](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/wrap-with-usememo.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+function isFunction(node) {
+    return t.isFunctionDeclaration(node) || t.isFunctionExpression(node) || t.isArrowFunctionExpression(node)
+}
+
+export function isVariableDeclarationWithNonFunctionInit(text) {
+    try {
+        const ast = codeToAst(text);
+        if (ast.program.body.length > 1) return false;
+        return t.isVariableDeclaration(ast.program.body[0]) && !isFunction((ast.program.body[0] as t.VariableDeclaration).declarations[0].init);
```
Comment (loonydev) [inline src/modules/wrap-with-usememo.ts:19]: Codacy Issue found: ['magic numbers' are not allowed: 0](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/rename-state.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+export async function renameState() {
+    const selectedStateVariable = selectedText();
+    const varName = await showInputBox(null, 'Name of the state variable');
+
+    const allAST = codeToAst(allText());
+    const containerPath = findPathInContext(allAST, selectedStateVariable);
+    const variableDeclarationParent = containerPath.findParent(parent => t.isVariableDeclaration(parent));
+    if (variableDeclarationParent && variableDeclarationParent.node.declarations[0].init.callee.name === 'useState') {
+        containerPath.scope.bindings[selectedStateVariable].path.parentPath.get('declarations.0.id.elements.0').scope.rename(selectedStateVariable, varName);
+        containerPath.scope.bindings[varName].path.parentPath.get('declarations.0.id.elements.1').scope.rename(`set${capitalizeFirstLetter(selectedStateVariable)}`, `set${capitalizeFirstLetter(varName)}`);
```
Comment (loonydev) [inline src/modules/rename-state.ts:41]: Codacy Issue found: [Exceeds maximum line length of 120](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/wrap-with-useeffect.ts [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+        const allAST = codeToAst(allText());
+        const containerPath = findPathInContext(allAST, text);
+        return !!containerPath.findParent(parent => {
+            return t.isFunctionExpression(parent)
+        });
+    } catch (e) {
+        return false;
+    }
+}
+export async function wrapWithUseEffect() {
```
Comment (loonydev) [inline src/modules/wrap-with-useeffect.ts:21]: Codacy Issue found: [expected call-signature: 'wrapWithUseEffect' to have a typedef](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/wrap-with-usememo.ts [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+import * as t from '@babel/types';
+import { pathContains } from "../ast-helpers";
+import traverse from "@babel/traverse";
+
+
+function isFunction(node) {
+    return t.isFunctionDeclaration(node) || t.isFunctionExpression(node) || t.isArrowFunctionExpression(node)
+}
+
+export function isVariableDeclarationWithNonFunctionInit(text) {
```
Comment (loonydev) [inline src/modules/wrap-with-usememo.ts:15]: Codacy Issue found: [expected parameter: 'text' to have a typedef](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/wrap-with-usememo.ts [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+        return false;
+    }
+}
+
+export async function wrapWithUseMemo() {
+    const snippet = selectedText();
+    const ast = codeToAst(snippet);
+    const memoWrapper = buildUseMemo({
+        VAR: (ast.program.body[0] as t.VariableDeclaration).declarations[0].id,
+        EXPRESSION: (ast.program.body[0] as t.VariableDeclaration).declarations[0].init
```
Comment (loonydev) [inline src/modules/wrap-with-usememo.ts:30]: Codacy Issue found: [Missing trailing comma](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/ast-helpers.ts (hunk 5 above)
Comment (loonydev) [inline src/ast-helpers.ts:17]: Codacy Issue found: [non-arrow functions are forbidden](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/wrap-with-usecallback.ts [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+function isFunction(path) {
+    return (t.isFunctionDeclaration(path) || isFunctionExpression(path))
+}
+
+
+export function findSelectedFunctionInRange(ast, start, end) {
+    let foundPath = null;
+    const visitor = {
+        'VariableDeclaration|FunctionDeclaration'(path) {
```
Comment (loonydev) [inline src/modules/wrap-with-usecallback.ts:23]: Codacy Issue found: [expected parameter: 'path' to have a typedef](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/wrap-with-usecallback.ts [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+import * as t from '@babel/types';
+import { pathContains } from "../ast-helpers";
+import traverse from "@babel/traverse";
+
+
+function isFunctionExpression(path) {
+    return t.isVariableDeclaration(path) || t.isFunctionExpression(path.node.init)
+}
+
+function isFunction(path) {
```
Comment (loonydev) [inline src/modules/wrap-with-usecallback.ts:15]: Codacy Issue found: [non-arrow functions are forbidden](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/settings.ts [hunk 19]
```diff
@@ -11,7 +13,7 @@ export const commonJSModuleSystemUsed = () => config().jsModuleSystem === 'commo
 
 const isExperimentOn = (experiment) => (config().experiments || []).includes(experiment);
 
-export const isHooksForFunctionalComponentsExperimentOn = () => isExperimentOn('hooksForFunctionalComponents');
+export const hooksSupported = () => ReactVersion[0] >= 16 && ReactVersion[1] >= 8;
```
Comment (loonydev) [inline src/settings.ts:16]: Codacy Issue found: ['magic numbers' are not allowed: 8](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/ast-helpers.ts [hunk 20]
```diff
@@ -11,3 +13,29 @@ export function getReactImportReference(ast): t.ImportDeclaration {
...
+        foundPath = path;
+      }
+    }
+  };
+
+  traverse(ast, visitor);
+  return foundPath;
+}
+
+export function pathContains(path, start, end) {
```
Comment (loonydev) [inline src/ast-helpers.ts:31]: Codacy Issue found: [expected parameter: 'end' to have a typedef](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/wrap-with-usecallback.ts [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+        return false;
+    }
+}
+
+function functionDeclartionToExpression(declarations: t.FunctionDeclaration) {
+    const { params, body, generator, async } = declarations
+    return t.functionExpression(null, params, body, generator, async)
+}
+
+export async function wrapWithUseCallback() {
```
Comment (loonydev) [inline src/modules/wrap-with-usecallback.ts:49]: Codacy Issue found: [non-arrow functions are forbidden](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/wrap-with-usememo.ts (hunk 12 above)
Comment (loonydev) [inline src/modules/wrap-with-usememo.ts:19]: Codacy Issue found: [Exceeds maximum line length of 120](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/wrap-with-usecallback.ts [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+function isFunctionExpression(path) {
+    return t.isVariableDeclaration(path) || t.isFunctionExpression(path.node.init)
+}
+
+function isFunction(path) {
+    return (t.isFunctionDeclaration(path) || isFunctionExpression(path))
+}
+
+
+export function findSelectedFunctionInRange(ast, start, end) {
```
Comment (loonydev) [inline src/modules/wrap-with-usecallback.ts:20]: Codacy Issue found: [expected parameter: 'end' to have a typedef](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

File: src/modules/wrap-with-usecallback.ts [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+import { replaceSelectionWith } from "../code-actions";
+import { buildEffectHook, buildUseCallbackHook } from "../snippet-builder";
+import { codeToAst, astToCode } from "../parsing";
+import { persistFileSystemChanges } from "../file-system";
+import * as t from '@babel/types';
+import { pathContains } from "../ast-helpers";
+import traverse from "@babel/traverse";
+
+
+function isFunctionExpression(path) {
```
Comment (loonydev) [inline src/modules/wrap-with-usecallback.ts:11]: Codacy Issue found: [non-arrow functions are forbidden](https://app.codacy.com/gh/wix/vscode-glean/pullRequest?prid=5535229)

**Comment by loonydev** [comment]
WTF, why codacy use my account -_-

**Comment by borislit** [comment]
@loonydev wondered about the same thing :) I was like "who is this dude critiquing me ? " :)

**Pull request merged**